File: swift/Sources/SdJwt/JwtTime.swift
import Foundation

public struct JwtTimeError: Error, CustomStringConvertible {
    public let description: String

    init(_ description: String) {
        self.description = description
    }
}

/// RFC 7519 time-claim validation (exp / nbf / iat) with configurable clock skew.
/// Fail-closed: malformed time claims are rejected, not ignored.
public struct JwtTimeValidator {
    private let now: () -> Date
    private let skewSeconds: Int64

    public init(now: @escaping () -> Date, skewSeconds: Int64 = 60) {
        self.now = now
        self.skewSeconds = skewSeconds
    }

    /// - Parameters:
    ///   - requireExp: reject tokens without `exp` (default: only validate if present)
    ///   - maxIatAgeSeconds: if set, reject tokens whose `iat` is older than this (freshness)
    public func validate(
        _ claims: JsonValue,
        requireExp: Bool = false,
        maxIatAgeSeconds: Int64? = nil
    ) throws {
        let nowSec = Int64(now().timeIntervalSince1970)

        if let exp = try numericDate(claims, "exp") {
            if nowSec > exp + skewSeconds {
                throw JwtTimeError("token expired (exp=\(exp), now=\(nowSec))")
            }
        } else if requireExp {
            throw JwtTimeError("missing required 'exp'")
        }

        if let nbf = try numericDate(claims, "nbf"), nowSec + skewSeconds < nbf {
            throw JwtTimeError("token not yet valid (nbf=\(nbf), now=\(nowSec))")
        }

        if let iat = try numericDate(claims, "iat") {
            if iat > nowSec + skewSeconds {
                throw JwtTimeError("iat is in the future (iat=\(iat), now=\(nowSec))")
            }
            if let maxAge = maxIatAgeSeconds, nowSec - iat > maxAge + skewSeconds {
                throw JwtTimeError("token too old (iat=\(iat), now=\(nowSec), max age=\(maxAge))")
            }
        } else if maxIatAgeSeconds != nil {
            throw JwtTimeError("freshness required but 'iat' missing")
        }
    }

    private func numericDate(_ claims: JsonValue, _ name: String) throws -> Int64? {
        switch claims[name] {
        case .none: return nil
        case let .numInt(n): return n
        case let .numDouble(d): return Int64(d)
        default: throw JwtTimeError("'\(name)' must be a number")
        }
    }
}

File: kotlin/sdjwt/src/main/kotlin/com/hopae/eudi/wallet/sdjwt/JwtTime.kt
package com.hopae.eudi.wallet.sdjwt

import java.time.Instant

class JwtTimeException(message: String) : Exception(message)

/**
 * RFC 7519 time-claim validation (exp / nbf / iat) with configurable clock skew.
 * Fail-closed: malformed time claims are rejected, not ignored.
 */
class JwtTimeValidator(
    private val now: () -> Instant,
    private val skewSeconds: Long = 60,
) {
    /**
     * @param requireExp reject tokens without `exp` (default: only validate if present)
     * @param maxIatAgeSeconds if set, reject tokens whose `iat` is older than this (freshness)
     */
    fun validate(
        claims: JsonValue.Obj,
        requireExp: Boolean = false,
        maxIatAgeSeconds: Long? = null,
    ) {
        val nowSec = now().epochSecond

        val exp = numericDate(claims, "exp")
        if (exp == null) {
            if (requireExp) throw JwtTimeException("missing required 'exp'")
        } else if (nowSec > exp + skewSeconds) {
            throw JwtTimeException("token expired (exp=$exp, now=$nowSec)")
        }

        numericDate(claims, "nbf")?.let { nbf ->
            if (nowSec + skewSeconds < nbf) throw JwtTimeException("token not yet valid (nbf=$nbf, now=$nowSec)")
        }

        val iat = numericDate(claims, "iat")
        if (iat != null) {
            if (iat > nowSec + skewSeconds) throw JwtTimeException("iat is in the future (iat=$iat, now=$nowSec)")
            if (maxIatAgeSeconds != null && nowSec - iat > maxIatAgeSeconds + skewSeconds) {
                throw JwtTimeException("token too old (iat=$iat, now=$nowSec, max age=$maxIatAgeSeconds)")
            }
        } else if (maxIatAgeSeconds != null) {
            throw JwtTimeException("freshness required but 'iat' missing")
        }
    }

    private fun numericDate(claims: JsonValue.Obj, name: String): Long? {
        return when (val v = claims[name]) {
            null -> null
            is JsonValue.NumInt -> v.value
            // NumericDate MAY be non-integer per RFC 7519; reject fractional-but-out-of-range
            is JsonValue.NumDouble -> v.value.toLong()
            else -> throw JwtTimeException("'$name' must be a number")
        }
    }
}
